src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -0,0 +1,109 @@
+
+        // The code below is equivalent to the commented out code above but has been tweaked
+        // to allow codegen to make some extra optimizations.
+
+        // The low byte of the packed result contains the hex representation of the incoming byte's low nibble.
+        // The adjacent byte of the packed result contains the hex representation of the incoming byte's high nibble.
+
+        // Finally, write to the output buffer starting with the *highest* index so that codegen can
+        // elide all but the first bounds check. (This only works if 'startingIndex' is a compile-time constant.)
+
+        // The JIT can elide bounds checks if 'startingIndex' is constant and if the caller is
+        // writing to a span of known length (or the caller has already checked the bounds of the
+        // furthest access).
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ToBytesBuffer(byte value, Span<byte> buffer, int startingIndex = 0, Casing casing = Casing.Upper)
+        {
+            uint difference = (((uint)value & 0xF0U) << 4) + ((uint)value & 0x0FU) - 0x8989U;
+            uint packedResult = ((((uint)(-(int)difference) & 0x7070U) >> 4) + difference + 0xB9B9U) | (uint)casing;
+
+            buffer[startingIndex + 1] = (byte)packedResult;
+            buffer[startingIndex] = (byte)(packedResult >> 8);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ToCharsBuffer(byte value, Span<char> buffer, int startingIndex = 0, Casing casing = Casing.Upper)
+        {
+            uint difference = (((uint)value & 0xF0U) << 4) + ((uint)value & 0x0FU) - 0x8989U;
+            uint packedResult = ((((uint)(-(int)difference) & 0x7070U) >> 4) + difference + 0xB9B9U) | (uint)casing;
+
+            buffer[startingIndex + 1] = (char)(packedResult & 0xFF);
+            buffer[startingIndex] = (char)(packedResult >> 8);
+        }
+
+#if NETCOREAPP || NETSTANDARD2_1
+        public static unsafe string ToString(ReadOnlySpan<byte> bytes, Casing casing = Casing.Upper)
+        {
+            fixed (byte* bytesPtr = bytes)
+            {
+                return string.Create(bytes.Length * 2, (Ptr: (IntPtr)bytesPtr, bytes.Length, casing), (chars, args) =>
+                {
+                    int p = 0;
+                    foreach (byte b in new ReadOnlySpan<byte>((byte*)args.Ptr, args.Length))
+                    {
+                        ToCharsBuffer(b, chars, p, args.casing);
+                        p += 2;
+                    }
+                });
+            }
+        }
+#endif
+
+        public static char ToCharUpper(int value)
+        {
+            value &= 0xF;
+            return (char)(value > 9 ? value - 10 + 'A' : value + '0');
+        }
+
+        public static char ToCharLower(int value)
+        {
+            value &= 0xF;
+            return (char)(value > 9 ? value - 10 + 'a' : value + '0');
+        }
```

Comment:
Did you measure this and find it to be faster than:
```C#
private static ReadOnlySpan<byte> HexEncodeUpper => new byte [] { (byte)'0', (byte)'1', (byte)'2', (byte)'3', (byte)'4', (byte)'5', (byte)'6', (byte)'7', (byte)'8', (byte)'9', (byte)'A', (byte)'B', (byte)'C', (byte)'D', (byte)'E', (byte)'F' };
private static ReadOnlySpan<byte> HexEncodeLower => new byte [] { (byte)'0', (byte)'1', (byte)'2', (byte)'3', (byte)'4', (byte)'5', (byte)'6', (byte)'7', (byte)'8', (byte)'9', (byte)'a', (byte)'b', (byte)'c', (byte)'d', (byte)'e', (byte)'f' };

public static char ToCharUpper(int value) => (char)HexEncodeUpper[value & 0xF];
public static char ToCharLower(int value) => (char)HexEncodeLower[value & 0xF];
```
? (Especially once the JIT properly removes the bounds check on this, if it doesn't already.)

Reply:
(Or using the same tricks/match employed by ToCharsBuffer above.)

Reply:
@stephentoub 

> Especially once the JIT properly removes the bounds check on this, if it doesn't already.

It doesn't today but I have a fix for JIT: https://github.com/EgorBo/runtime-1/commit/bcc095eb01fce5c69b075761950979d63099c0a1
so it will remove them for things like `array[i % c]` or `array[i & c]` if `c` is less than `array.Length`.

Reply:
> Did you measure this and find it to be faster than:
I didn't as these are convenience methods rarely called more than a few times. The Span versions are intended for perf sensitive code.

Also, introducing global memory blob in each assembly which uses this code seems to me contra-productive size-wise.

Reply:
> Did you measure this and find it to be faster than:

Here are the micro-benchmarks for 3 different cases. The sources for them at https://gist.github.com/marek-safar/b26e2bfc9a60a2bd348b3f36d3cb8288

```
BenchmarkDotNet=v0.12.0, OS=macOS 10.15.2 (19C57) [Darwin 19.2.0]
Intel Core i7-7920HQ CPU 3.10GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-alpha.1.20061.5
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.6301, CoreFX 4.700.19.56404), X64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.6007, CoreFX 5.0.20.6007), X64 RyuJIT

|           Method |      Mean |     Error |    StdDev |
|----------------- |----------:|----------:|----------:|
|   Test_Condition | 0.0257 ns | 0.0310 ns | 0.0290 ns |
|      Test_String | 0.1618 ns | 0.0320 ns | 0.0299 ns |
|         Test_ROS | 0.5877 ns | 0.0473 ns | 0.0506 ns |
```

Reply:
I don't think that benchmark is testing what you think it's testing.

Convert_Condition can be inlined in your test, and once it is, constant propagation results in Test_Condition compiling down to:
```
    L0000: mov eax, 0x45
    L0005: movzx eax, ax
    L0008: ret
```

In contrast, if I change the benchmark to instead be:
```C#
    public int Test_Condition()
    {
        int sum = 0;
        for (int i = 0; i < 16; i++)
            sum += Convert_Condition(i);
        return sum;
    }

    public int Test_String()
    {
        int sum = 0;
        for (int i = 0; i < 16; i++)
            sum += Convert_String(i);
        return sum;
    }
```
then Convert_Condition doesn't get inlined, Convert_String does get inlined, and I get these numbers:

|         Method |      Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |----------:|----------:|----------:|------:|------:|------:|----------:|
| Test_Condition | 30.822 ns | 0.1755 ns | 0.1555 ns |     - |     - |     - |         - |
|    Test_String |  7.531 ns | 0.0119 ns | 0.0106 ns |     - |     - |     - |         - |

Reply:
These are too short to give us a good indication. I wrote a bit more realistic one and conditional one is still the winner https://gist.github.com/marek-safar/670c40f810cfd09b1498c52f1f8b6978

Reply:
I commented on the gist as well, but this greatly depends on the machine you are running against, the input data, and a number of other factors.

On my box (I can run on several others as well):
```
BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET Core SDK=5.0.100-alpha.1.20064.2
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.6307, CoreFX 5.0.20.6307), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.6307, CoreFX 5.0.20.6307), X64 RyuJIT


|           Method |                Input |     Mean |    Error |   StdDev |
|----------------- |--------------------- |---------:|---------:|---------:|
|    StringVersion | 18446744073709551615 | 44.64 ns | 0.235 ns | 0.220 ns |
| ConditionVersion | 18446744073709551615 | 58.82 ns | 1.199 ns | 1.516 ns |
|       ROSVersion | 18446744073709551615 | 43.57 ns | 0.291 ns | 0.227 ns |
```

The main issue with the "Condition" version is that it involves two branches and its predictability depends greatly on the input alternating hex and numeric digits is basically the worst case scenario.

It could likely be improved by changing to a single branch, such as the following:
```
        value &= 0xF;
        value += '0';
        
        if (value > '9')
        {
            value += ('a' - '0');
        }
        
        return (char)value;
```
In this scenario, the branch predictor has less to worry about and we still aren't doing any extra work.

An even better approach, especially since the inputs will likely always be fairly unpredictable would be to use `cmov` and avoid branching altogether. This is what GCC, Clang, and MSVC do; as an example (but is namely only beneficial for inputs that can't be accurately predicted). However, I don't believe the JIT has support for this today.

Reply:
> It could likely be improved by changing to a single branch

I applied that and re-ran the benchmarks. I see improvement in some of the test cases with the change you recommended (that reduces branches). If we are going with this change already, we should make the change @tannergooding recommended.

Reply:
I can change that it has a minimal negative impact locally.

@tannergooding I'd interested in numbers from arm64 if you can run it there

Reply:
> I'd interested in numbers from arm64 if you can run it there

Certainly. I'm doing a build for this now and should be able to test sometime after lunch.

Reply:
From my Samsung Galaxy Book 2:
```
BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Snapdragon 850 2.96 GHz, 1 CPU, 8 logical and 8 physical cores
.NET Core SDK=5.0.100-alpha1-015515
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.19.52501, CoreFX 5.0.19.52503), Arm64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.19.52501, CoreFX 5.0.19.52503), Arm64 RyuJIT


|           Method |                Input |     Mean |   Error |  StdDev |
|----------------- |--------------------- |---------:|--------:|--------:|
|    StringVersion | 18446744073709551615 | 147.3 ns | 0.78 ns | 0.73 ns |
| ConditionVersion | 18446744073709551615 | 125.9 ns | 0.32 ns | 0.28 ns |
|       ROSVersion | 18446744073709551615 | 119.1 ns | 0.40 ns | 0.35 ns |
```

Reply:
Is the `ConditionVersion` in this case the one with the 2 branches or the one with your suggestion (and only 1 branch)?

Reply:
Its thr code from @Marek-safar's gist, with no changes

Reply:
How does the single branch condition perform on `Samsung Galaxy Book 2` in comparison to the rest? If it is on-par/better than ROSVersion, then that helps answer which approach we should take (granted, with @EgorBo's changes, the ROSVersion could get better).